### PR TITLE
Capture The Flag: Class Warfare

### PR DESCRIPTION
**Pull request by Thunder12345**
## About The Pull Request

Class Warfare brings a major new expansion to CTF gameplay, with multiple classes providing options for the player's role in the team and style of combat.

### The Classes:

**Rifleman:** The current CTF loadout. A decent all aspect class capable in both offensive and defensive fighting.

**Assaulter:** Offensive class. Armed with a short range shotgun and rocket gloves, very weak shield.

**Marksman:** Sniper class. Armed with a reduced damage hitscan rifle.

Stats and balance details can be found here: https://hackmd.io/@Thunder12345/rJjEq4Gou

- [x] Move player equipment and outfits out of the main ctf game code
- [x] Rewrite multi-class spawning code to use radial menus
- [x] Rewrite existing equipment code to remove snowflakes
- [x] Develop new classes
- [x] Implement gear and outfits for new classes
- [x] Produce icons for new gear and outfits
- [x] Final polishing pass
- [x] Refactor medieval simulation 
- [x] Playtest for balance and bugfixing

## Why It's Good For The Game

Current CTF gameplay has limited depth, as no matter the choice of map everyone runs around spamming bullets at each other. By giving players a more diverse class choice more depth is available through interactions and balancing between classes.

## Changelog
:cl:
add: Added class mechanics to Capture The Flag, with two new classes.
add: The Assaulter class is a fast skirmisher armed with a shotgun and rocket gloves, but a very fragile energy shield.
add: The Marksman class is a sniper armed with a hitscan rifle doing lower than standard damage.
imageadd: Added a bunch of icons for Assaulter and Marksman gear.
tweak: The fair maidens on the medieval simulation shuttle are no longer weapons of mass destruction
fix: You can no longer spawn outside of CTF matches when multiple classes are available.
refactor: Rewritten the multi-class part of CTF spawning to use radial menus.
refactor: Medieval simulation has been refactored to bring it fully under the CTF framework
/:cl:

### CODE REVIEW

**Comment by Fikou** [comment]
refactor

**Comment by optimumtact** [comment]
this is a secondary admin tool and you're bloating it incomprehensibly.

The more complex something is and the less it gets used the less it will work.

**Comment by Fikou** [comment]
> this is a secondary admin tool and you're bloating it incomprehensibly.
> 
> The more complex something is and the less it gets used the less it will work, so all these changes to CTF are just going to result in a broken tool that will eventually get removed.

Welcome to /tg/station13 CTF.
After nine years in development, hopefully it will have been worth the wait.
Thanks, and have fun.

File: code/datums/elements/delete_on_drop.dm [hunk 1]
```diff
@@ -0,0 +1,20 @@
+/**
+ * Attaches to an item, if that item is dropped on the floor delete it
+ */
+/datum/element/delete_on_drop
```
Comment (Fikou) [inline code/datums/elements/delete_on_drop.dm:4]: why did you add this theres an item flag that does exactly the same thing

Comment (Thunder12345) [reply code/datums/elements/delete_on_drop.dm:4]: There is?

**Comment by YakumoChen** [comment]
Is CTF really an admin tool? Players can activate it in-round at any time....

**Comment by Timberpoes** [comment]
PR author requested removal of stale label as the PR is still actively being worked on.

**Comment by Redrover1760** [comment]
I like it. Gives people the ability to improve their skills at some more underused aspects of the game outside of just lasers. Shotguns, rocket gloves, and snipers? Nice.

**Comment by adamkad1** [comment]
Medic and engineer classes when?

**Review by Twaticus — APPROVED** [review]

**Review by Krysonism — APPROVED** [review]
I think the sprites looks pretty good. The design of the CTF shotgun is a little bit boring, but i think that is fine since CTF is a side activity.

File: code/datums/elements/delete_on_drop.dm [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+	var/list/myvar = list()
+
+/datum/element/delete_on_drop/Attach(datum/target)
+	. = ..()
+	if(!isitem(target))
+		return COMPONENT_INCOMPATIBLE
+	RegisterSignal(target, COMSIG_ITEM_DROPPED, .proc/del_on_drop)
+
+/datum/element/delete_on_drop/Detach(datum/source)
+	UnregisterSignal(source, COMSIG_ITEM_DROPPED)
```
Comment (Kylerace) [inline code/datums/elements/delete_on_drop.dm:15]: you dont call parent here

File: code/modules/capture_the_flag/ctf_classes.dm [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+	// Make clothing in the specified slots NODROP
+	for(var/slot in nodrop_slots)
+		no_drops += H.get_item_by_slot(slot)
+	// Make items in the hands NODROP
+	for(var/obj/item/I in H.held_items)
+		no_drops += I
+	listclearnulls(no_drops) // For any slots we didn't have filled
+	// Apply TRAIT_NODROP to everything
+	for(var/i in no_drops)
+		var/obj/item/I = i
```
Comment (Kylerace) [inline code/modules/capture_the_flag/ctf_classes.dm:53]: use as anything here, also name

File: code/modules/capture_the_flag/ctf_classes.dm [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+	///Icon state for this class
+	var/icon_state = "ctf_rifleman"
+	///Do they get a headset?
+	var/has_radio = TRUE
+	///Do they get an ID?
+	var/has_card = TRUE
+	///Which slots to apply TRAIT_NODROP to the items in
+	var/list/nodrop_slots = list(ITEM_SLOT_OCLOTHING, ITEM_SLOT_GLOVES, ITEM_SLOT_FEET, ITEM_SLOT_ICLOTHING, ITEM_SLOT_EARS)
+
+/datum/outfit/ctf/post_equip(mob/living/carbon/human/H, visualsOnly=FALSE)
```
Comment (Kylerace) [inline code/modules/capture_the_flag/ctf_classes.dm:32]: please make this a descriptive variable as single letter variables are against our contributing guidelines and are cringe

File: code/modules/capture_the_flag/ctf_classes.dm [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+	if(visualsOnly)
+		return
+	var/list/no_drops = list()
+
+	if(has_card)
+		var/obj/item/card/id/W = H.wear_id
```
Comment (Kylerace) [inline code/modules/capture_the_flag/ctf_classes.dm:38]: descriptive name please

File: code/modules/capture_the_flag/ctf_classes.dm [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+		no_drops += W
+		W.registered_name = H.real_name
+		W.update_label()
+		W.update_icon()
+
+	// Make clothing in the specified slots NODROP
+	for(var/slot in nodrop_slots)
+		no_drops += H.get_item_by_slot(slot)
+	// Make items in the hands NODROP
+	for(var/obj/item/I in H.held_items)
```
Comment (Kylerace) [inline code/modules/capture_the_flag/ctf_classes.dm:48]: descriptive name please

File: code/modules/capture_the_flag/ctf_classes.dm [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+		no_drops += I
+	listclearnulls(no_drops) // For any slots we didn't have filled
+	// Apply TRAIT_NODROP to everything
+	for(var/i in no_drops)
+		var/obj/item/I = i
+		ADD_TRAIT(I, TRAIT_NODROP, CAPTURE_THE_FLAG_TRAIT)
+
+	if(has_radio)
+		var/obj/item/radio/R = H.ears
```
Comment (Kylerace) [inline code/modules/capture_the_flag/ctf_classes.dm:57]: name

File: code/modules/capture_the_flag/ctf_equipment.dm [hunk 8]
```diff
@@ -0,0 +1,605 @@
...
+		damage = 0
+		return PROJECTILE_PIERCE_NONE /// hey uhhh don't hit anyone behind them
+	. = ..()
+
+/obj/projectile/beam/ctf/on_hit(atom/target, blocked = FALSE)
+	. = ..()
+	if(is_ctf_target(target) && blocked == FALSE)
+		if(iscarbon(target))
+			var/mob/living/carbon/M = target
+			M.adjustBruteLoss(150, 0)
```
Comment (Kylerace) [inline code/modules/capture_the_flag/ctf_equipment.dm:19]: name

File: code/modules/capture_the_flag/ctf_equipment.dm [hunk 9]
```diff
@@ -0,0 +1,605 @@
...
+	icon_state = "purple_laser"
+	damage = 200
+	damage_type = BURN
+	impact_effect_type = /obj/effect/temp_visual/impact_effect/purple_laser
+	light_color = LIGHT_COLOR_PURPLE
+
+/obj/projectile/beam/instakill/on_hit(atom/target)
+	. = ..()
+	if(iscarbon(target))
+		var/mob/living/carbon/M = target
```
Comment (Kylerace) [inline code/modules/capture_the_flag/ctf_equipment.dm:188]: name

**Review by Kylerace — CHANGES_REQUESTED** [review]
code seems ok at 4:30 am, main problem is var names

**Comment by Stalkeros2** [comment]
Based as hell.
Medics and engineers when.
